File: problems/12.heaps/ts/MinHeap.ts
export class MinHeap {
    private heap: number[] = [];

    constructor(initialHeap: number[]) {
        this.heap = initialHeap;
        this.buildHeap();
    }

    extractMin() {
        if (this.heap.length === 0) {
            throw new Error('Heap is empty');
        }

        const min = this.heap[0];
        this.heap[0] = this.heap.pop() as number;
        this.heapifyDown(0);
        return min;
    }

    insert(value: number) {
        this.heap.push(value);
        this.heapifyUp(this.heap.length - 1);
    }

    getHeap() {
        return this.heap;
    }

    private buildHeap() {
        for (let i = Math.floor((this.heap.length - 1) / 2); i >= 0; --i) {
            this.heapifyDown(i);
        }
    }

    private heapifyUp(index: number) {
        if (index === 0) return;
        let parentIndex = Math.floor((index - 1) / 2); // 0

        if (this.heap[parentIndex] > this.heap[index]) {
            let temp = this.heap[parentIndex];
            this.heap[parentIndex] = this.heap[index];
            this.heap[index] = temp;
            this.heapifyUp(parentIndex);
        }
    }

    private heapifyDown(index: number) {
        let leftIndex = index * 2 + 1;
        let rightIndex = index * 2 + 2;
        let left = this.heap[leftIndex];
        let right = this.heap[rightIndex];

        let minChildIndex = left < right ? leftIndex : rightIndex;

        if (this.heap[index] > this.heap[minChildIndex]) {
            let temp = this.heap[index];
            this.heap[index] = this.heap[minChildIndex];
            this.heap[minChildIndex] = temp;
            this.heapifyDown(minChildIndex);
        }
    }
}

File: golang/12.heaps/min_heap.go
package main

import (
	"math"
	"reflect"
)

type MinHeap struct {
	heap []int
}

func NewMinHeap(initialHeap []int) MinHeap {
	maxHeap := MinHeap{
		heap: []int{},
	}
	maxHeap.buildHeap(initialHeap)

	return maxHeap
}

func (h MinHeap) buildHeap(initialHeap []int) {
	for _, value := range initialHeap {
		h.heap = append(h.heap, value)
		h.heapifyUp(len(h.heap) - 1)
	}
}

func (h MinHeap) Insert(value int) {
	h.heap = append(h.heap, value)
	h.heapifyUp(len(h.heap) - 1)
}

func (h MinHeap) Peek() int {
	if len(h.heap) == 0 {
		return math.MinInt32
	}

	max := h.heap[0]
	h.heap = h.heap[1:]
	swapper := reflect.Swapper(h.heap)
	swapper(0, len(h.heap)-1)
	h.heapifyDown(0)

	return max
}

func (h MinHeap) GetHeap() []int {
	return h.heap
}

func (h MinHeap) heapifyDown(parentIndex int) {
	leftChildIndex := 2*parentIndex + 1
	rightChildIndex := 2*parentIndex + 2

	targetIndex := leftChildIndex

	if rightChildIndex < len(h.heap) && leftChildIndex < len(h.heap) {
		if h.heap[leftChildIndex] < h.heap[rightChildIndex] {
			targetIndex = rightChildIndex
		}
	}

	if targetIndex < len(h.heap) {
		if h.heap[targetIndex] > h.heap[parentIndex] {
			swapper := reflect.Swapper(h.heap)
			swapper(targetIndex, parentIndex)
			h.heapifyDown(targetIndex)
		}
	}
}

func (h MinHeap) heapifyUp(childIndex int) {
	parentIndex := int(math.Floor(float64(childIndex) / 2))

	if parentIndex > 0 && parentIndex < len(h.heap) {
		if h.heap[parentIndex] < h.heap[childIndex] {
			swapper := reflect.Swapper(h.heap)
			swapper(parentIndex, childIndex)

			h.heapifyUp(parentIndex)
		}
	}
}
